fe/fe-core/src/main/java/org/apache/doris/planner/PartitionSortNode.java

```diff
@@ -0,0 +1,228 @@
+        }
+        output.append("\n");
+
+        // Add the limit information;
+        output.append(prefix).append("has global limit: ").append(hasGlobalLimit).append("\n");
+        output.append(prefix).append("partition limit: ").append(partitionLimit).append("\n");
+
+        return output.toString();
+    }
+
+    private void removeUnusedExprs() {
+        if (!isUnusedExprRemoved) {
+            if (resolvedTupleExprs != null) {
+                List<SlotDescriptor> slotDescriptorList = this.info.getSortTupleDescriptor().getSlots();
+                for (int i = slotDescriptorList.size() - 1; i >= 0; i--) {
+                    if (!slotDescriptorList.get(i).isMaterialized()) {
+                        resolvedTupleExprs.remove(i);
+                        nullabilityChangedFlags.remove(i);
+                    }
+                }
+            }
+            isUnusedExprRemoved = true;
+        }
+    }
+
+    @Override
+    protected void toThrift(TPlanNode msg) {
+        msg.node_type = TPlanNodeType.PARTITION_SORT_NODE;
+
+        TSortInfo sortInfo = info.toThrift();
+        Preconditions.checkState(tupleIds.size() == 1, "Incorrect size for tupleIds in PartitionSortNode");
+        removeUnusedExprs();
+        if (resolvedTupleExprs != null) {
+            sortInfo.setSortTupleSlotExprs(Expr.treesToThrift(resolvedTupleExprs));
+            // FIXME this is a bottom line solution for wrong nullability of resolvedTupleExprs
+            // remove the following line after nereids online
+            sortInfo.setSlotExprsNullabilityChangedFlags(nullabilityChangedFlags);
+        }
+
+        TopNAlgorithm topNAlgorithm;
+        if (function == WindowFuncType.ROW_NUMBER) {
+            topNAlgorithm = TopNAlgorithm.ROW_NUMBER;
+        } else if (function == WindowFuncType.RANK) {
+            topNAlgorithm = TopNAlgorithm.RANK;
+        } else {
+            topNAlgorithm = TopNAlgorithm.DENSE_RANK;
+        }
+
+        TPartitionSortNode partitionSortNode = new TPartitionSortNode();
+        partitionSortNode.setTopNAlgorithm(topNAlgorithm);
+        partitionSortNode.setPartitionExprs(Expr.treesToThrift(partitionExprs));
+        partitionSortNode.setSortInfo(sortInfo);
+        partitionSortNode.setHasGlobalLimit(hasGlobalLimit);
+        partitionSortNode.setPartitionInnerLimit(partitionLimit);
+        msg.partition_sort_node = partitionSortNode;
+    }
+
+    @Override
+    public Set<SlotId> computeInputSlotIds(Analyzer analyzer) throws NotImplementedException {
+        removeUnusedExprs();
+        List<Expr> materializedTupleExprs = new ArrayList<>(resolvedTupleExprs);
+        List<SlotId> result = Lists.newArrayList();
+        Expr.getIds(materializedTupleExprs, null, result);
+        return new HashSet<>(result);
+    }
+
+    /**
+     * Supplement the information needed by be for the partition sort node.
+     */
+    public void finalizeForNereids(TupleDescriptor tupleDescriptor,
```

Comment:
could we put all things into constructor? `finalizeForNereids` will be removed in the future, new Node should not implement this function